tos/chips/cortex/m4/sam4/l/timers/HalSam4lASTP.nc

```diff
@@ -87,52 +85,43 @@ implementation
     }
     async command void Alarm.startAt(uint32_t t0, uint32_t dt)
     {
+        /* IMPLEMENTATION DETAILS
+         * The AST is a 16kHz clock that wraps at 32 bits. We want to provide the abstraction of a
+         * 32 kHz clock that wraps at 32 bits, or equivalently a 16 kHz clock that wraps at 31 bits.
+         * We can model the AST as two clocks:
+         * 1) A short clock that operates at 16 kHz and wraps at 31 bits
+         * 2) A long clock that operates at 16 kHz and wraps at 32 bits
+         * The code below calculates the counter value on the AST on the long clock, and calls the
+         * value cv. It converts t0 to a value on the short clock. Then it computes t1 = t0 + dt on
+         * the long clock.
+         */
+
         uint32_t cv, t1;
-        call ast.disable();
-        call ast.disableAlarmIRQ();
-        call ast.clearAlarm();
-        cv = call ast.getCounterValue();
-        
         atomic {
-            if (cv > (uint32_t) 0x7FFFFFFF) { // artificially overflow at 31 bits
-                cv &= (uint32_t) 0x7FFFFFFF;
+            call ast.disable();
+            call ast.disableAlarmIRQ();
+            call ast.clearAlarm();
+            
+            cv = (call ast.getCounterValue()) & (uint32_t) 0x7FFFFFFF;
+            t1 = (t0 >> 1) + (dt >> 1) + (t0 & dt & 1);
+            t0 >>= 1;
+             
+            if (t0 > cv) {
+                cv |= (uint32_t) 0x80000000;
+            }
+            
+            if (t1 <= cv) {
+                // Fire the alarm right away
+                call ast.enable();
+                signal Alarm.fired();
+            } else {
+                // Wait for t1 - cv ticks
                 call ast.setCounterValue(cv);
+                call ast.setAlarm(t1);
```

Comment:
The fix in getAlarm above is invalid. If you are concerned that setting the alarm is causing a race, the correct solution is to block until the set completes. I am not sure if the registers will sync while the clock is stopped, however. You should try and see if they will by modifying the setAlarm() function to include a while(busy) spin just before it returns. If the chip hangs indefinitely, then we know it doesn't progress while the AST is stopped. If it does not hang, then that will solve not only races with getAlarm() but also races with the AR==CV interrupt. As it stands it might be possible for CV to tick past AR before the register syncs